sentry/src/main/java/io/sentry/featureflags/FeatureFlagBuffer.java

```diff
@@ -0,0 +1,207 @@
+package io.sentry.featureflags;
+
+import io.sentry.ISentryLifecycleToken;
+import io.sentry.ScopeType;
+import io.sentry.SentryOptions;
+import io.sentry.protocol.FeatureFlag;
+import io.sentry.protocol.FeatureFlags;
+import io.sentry.util.AutoClosableReentrantLock;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+@ApiStatus.Internal
+public final class FeatureFlagBuffer implements IFeatureFlagBuffer {
+
+  private volatile @NotNull CopyOnWriteArrayList<FeatureFlagEntry> flags;
+  private final @NotNull AutoClosableReentrantLock lock = new AutoClosableReentrantLock();
+  private int maxSize;
+
+  private FeatureFlagBuffer(int maxSize) {
+    this.maxSize = maxSize;
+    this.flags = new CopyOnWriteArrayList<>();
+  }
+
+  private FeatureFlagBuffer(
+      int maxSize, final @NotNull CopyOnWriteArrayList<FeatureFlagEntry> flags) {
+    this.maxSize = maxSize;
+    this.flags = flags;
+  }
+
+  private FeatureFlagBuffer(@NotNull FeatureFlagBuffer other) {
+    this.maxSize = other.maxSize;
+    this.flags = other.flags;
+  }
+
+  @Override
+  public void add(@NotNull String flag, boolean result) {
+    try (final @NotNull ISentryLifecycleToken ignored = lock.acquire()) {
+      final int size = flags.size();
+      final @NotNull ArrayList<FeatureFlagEntry> tmpList = new ArrayList<>(size + 1);
+      for (FeatureFlagEntry entry : flags) {
+        if (!entry.flag.equals(flag)) {
+          tmpList.add(entry);
+        }
+      }
+      tmpList.add(new FeatureFlagEntry(flag, result, System.nanoTime()));
+
+      if (tmpList.size() > maxSize) {
+        tmpList.remove(0);
+      }
+
+      flags = new CopyOnWriteArrayList<>(tmpList);
```

Comment:
```suggestion
      final int size = flags.size();
      for (int i = 0; i < size; i++) {
        if (flags.get(i).equals(flag)) {
          flags.remove(i);
          break;
        }
      }
      flags.add(new FeatureFlagEntry(flag, result, System.nanoTime()));

      if (flags.size() > maxSize) {
        flags.remove(0);
      }
```
Maybe this is better?